Month 7/step 1

frrenzy opened the pull request:
Промежуточная проверка проекта

frrenzy committed:
feat: add AppHeader markup

frrenzy committed:
refactor: decompose AppHeader component

frrenzy committed:
feat: add BurgerIngredients component

frrenzy committed:
fix(config): fix stylelint warning on scrollbar pseudo-classes

frrenzy committed:
fix(props): add isRequired to shape

frrenzy committed:
refactor: reorganize imports

frrenzy committed:
feat: add ingredients cart

frrenzy committed:
feat: add BurgerConstructor component

frrenzy committed:
feat: add delete from cart action

frrenzy committed:
feat(docs): add README.md

frrenzy committed:
refactor: reorganize imports

frrenzy committed:
refactor: remove redundant styles

frrenzy committed:
fix(build): fix gh-pages config

frrenzy committed:
feat: add GH Pages deployment

Zaytsev-Alex commented on package.json:
```diff
@@ -14,7 +16,7 @@
     "eject": "react-scripts eject",
     "predeploy": "yarn build",
-    "deploy": "gh-pages -d dist",
+    "deploy": "gh-pages -d build",
```
Отлично: очень круто, что реализован деплой!

Zaytsev-Alex commented on src/components/header-button/header-button.jsx:
```diff
@@ -0,0 +1,36 @@
+  }
+
+  return (
+    <button
```
Можно лучше: сейчас в идеале сделать ссылки в хедере ссылками (элемент a), чтобы увеличить семантичность. Дабы избежать обновление страницы в ссылки можно передать символ решетки в качестве значения href. Далее ссылки будут заменены на компонент ссылки из библиотеки react-router.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/app-header/index.js:
```diff
@@ -0,0 +1 @@
+export { default } from './app-header'
```
Отлично, что добавлены index.js файлы для упрощения импортов! Это очень удобно.

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.module.scss:
```diff
@@ -0,0 +1,27 @@
+  display: flex;
+  justify-content: end;
+  align-items: center;
+  gap: calc(2 * var(--offset-base-size));
```
Можно лучше: когда курсор пользователя на карточке его можно менять на pointer. Так пользователь будет знать, что с карточкой можно взаимодействовать.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,43 @@
+  const [cart, setCart] = useState(initialCart.length > 0 ? initialCart : [])
+
+  const updateCart = _id => () => {
+    setCart(cart.filter(item => item._id !== _id))
```
Отлично, что уже сейчас реализована дополнительная функциональность, не требуемая по заданию.

Zaytsev-Alex commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,43 @@
+const App = () => {
+  const [cart, setCart] = useState(initialCart.length > 0 ? initialCart : [])
+
+  const updateCart = _id => () => {
```
Можно лучше: т.к. функция выполняет удаление ингредиента, ее стоит так и назвать. Имена функций должна отображать точно то, что они делают. updateCard слишком общее название в этом случае.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,43 @@
+import styles from './app.module.scss'
+
+const App = () => {
+  const [cart, setCart] = useState(initialCart.length > 0 ? initialCart : [])
```
Можно лучше: здесь и далее: если переменная хранит массив, ее принято называть во множественном числе: carts, setCarts, initialCarts.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,80 @@
+
+  const getTypeByIndex = useCallback(
+    idx =>
+      idx === 0
+        ? { type: 'top', isLocked: true }
+        : idx === cartLength - 1
+        ? { type: 'bottom', isLocked: true }
+        : { type: undefined, isLocked: undefined },
```
Можно лучше: вложенные тернарные операторы слишком сложны в понимании, лучше это заменить на несколько конструкций с использованием if.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,80 @@
+      </ul>
+      <div className={`${burgerConstructorStyles.controls} mt-10 mr-4`}>
+        <Price
+          value={cart.reduce((total, { price }) => total + price, 0)}
```
Отлично, что уже сейчас реализован какой-то подсчет суммы, а не используются захардкоженные данные.
Но, можно лучше: данные для рендера стоит готовить до рендера, это упрощает понимание проекта. Лучше создать переменную в теле компонента и в ней выполнить подсчет цены, а здесь ее передавать.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,80 @@
+
+BurgerConstructor.propTypes = {
+  cart: PropTypes.arrayOf(
+    PropTypes.shape({
+      _id: PropTypes.string.isRequired,
+      name: PropTypes.string.isRequired,
+      type: PropTypes.string.isRequired,
+      price: PropTypes.number.isRequired,
+      image: PropTypes.string.isRequired,
+      image_mobile: PropTypes.string.isRequired,
+      image_large: PropTypes.string.isRequired,
+      count: PropTypes.number.isRequired,
+    }),
```
Можно лучше: т.к. объект ингредиента является в приложении основной единицей данных, он может быть часто использован в качестве проперти компонента, следовательно его тип тоже. Стоит вынести определение типа в отдельный файл, к примеру utils/prop-types, там полностью описать тип ингредиента и переиспользовать его в компонентах.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,128 @@
+  const refs = useRef([])
+  const setRef = index => type => el => (refs.current[index] = { type, el })
+
+  const handleTabClick = value => {
+    const index = refs.current.findIndex(item => item.type === value)
+    refs.current[index].el.scrollIntoView({ behavior: 'smooth' })
+    setCurrent(value)
+  }
```
Отлично, что реализован скролл к выбранной категории!

Zaytsev-Alex commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,128 @@
+
+  const buns = {
+    name: 'Булки',
+    type: 'bun',
+    ingredients: ingredients.filter(ingredient => ingredient.type === 'bun'),
+    cart: cart.filter(ingredient => ingredient.type === 'bun'),
```
Можно лучше: типы ингредиентов будут часто использоваться и хорошим решением будет вынести типы в константу:
```
const INGREDIENT_TYPES = {
    BUN: 'bun',
    SAUCE: 'sauce',
    MAIN: 'main'
};
```

и использовать данные из нее. Это улучшит поддерживаемость кода и так же позволит избавиться от магических строк. https://ru.hexlet.io/blog/posts/magic-numbers Статья про магические числа, но тоже применимо и к строкам.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,128 @@
+  )
+}
+
+BurgerIngredients.propTypes = {
```
Можно лучше: вот здесь уже дублируется описание типа ингредиента, можно будет переиспользовать вынесенный тип.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/price/price.jsx:
```diff
@@ -0,0 +1,30 @@
+
+const Price = ({ value, size = 'default' }) => {
+  return (
+    <>
```
Можно лучше: стоит не использовать фрагменты, если в них нет необходимости. Они нужны, если надо отрендерить больше 1 элемента на верхнем уровне. В этом случае рендерится только один элемент параграфа.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/ingredients-section/ingredients-section.jsx:
```diff
@@ -0,0 +1,86 @@
+      <ul className={`${ingredientsSectionStyles.list} pt-6 pr-4 pl-4 pb-10`}>
+        {ingredients.length ? (
+          ingredientsWithCount.map(({ image, name, price, _id, count }) => (
+            <li key={_id}>
```
Отлично, что в этой секции приложения используется _id в качестве key. Это оптимальное решение.

Zaytsev-Alex commented on src/components/ingredient-card/ingredient-card.jsx:
```diff
@@ -0,0 +1,42 @@
+    <figure className={ingredientCardStyles.figure}>
+      <img
+        src={image}
+        alt={name}
```
Отлично, что для картинки используется альтернативный текст, так пользователи с программами чтения с экрана будут знать, что изображено на картинке.

Zaytsev-Alex commented on src/components/ingredient-card/ingredient-card.jsx:
```diff
@@ -0,0 +1,42 @@
+        alt={name}
+        className={`mr-4 ml-4 mb-1 ${ingredientCardStyles.image}`}
+      />
+      {count > 0 && (
+        <Counter
```
Отлично, что каунтер отображается только, если его значение больше 0. Так UI приложения будет лучше.

Zaytsev-Alex commented on src/components/ingredient-card/ingredient-card.jsx:
```diff
@@ -0,0 +1,42 @@
+  )
+}
+
+IngredientCard.propTypes = {
+  image: PropTypes.string.isRequired,
+  name: PropTypes.string.isRequired,
+  price: PropTypes.number.isRequired,
+  count: PropTypes.number.isRequired,
+}
```
Можно лучше: стоит передавать в карточку ингредиента полный объект ингредиента и здесь уже вычленять нужное. Если в новых требованиях понадобится, что-то еще отображать на карточке, надо будет менять это в нескольких местах, а не только в этом компоненте. Также это может потребоваться далее для реализации модальных окон.

frrenzy replied:
+

Zaytsev-Alex reviewed (commented):
Отличная работа, все требования предварительного этапа реализованы, приложение совпадает с макетом. Классно, сделан дополнительный функционал, не требуемый по заданию, и проект задеплоен. Применено много крутых решений.

Критических замечаний нет.

Из того, что можно сделать лучше, хочу отметить следующее:
- Имя переменной, должно отображать то, что она содержит. Если значение является массивом, имя стоит задавать во множественном числе. Имена функций также должны отражать, что они делают, по возможности максимально точно.
- Избегать вложенных тернарных оператор;
- Готовить данные для рендера до рендера;
- Вынести полную типизацию объекта ингредиента в отдельный файл и переиспользовать полученный тип;
- Вынести типы ингредиентов в отдельную константу;
- Сделать ссылки в хедера с использованием элемента ссылки a.

Больше и детальнее комментариев можно найти по ходу проверки.

Хочу напомнить, что работу можно будет принять, только после выполнения требований для финального этапа и исправления всех "Надо исправить" (сейчас их нет, но могу появится на финальном этапе).

Удачи в дальнейшем выполнении этой работы!

Commit:
refactor: review #1

Убрал стейт корзины, подумал, что незачем хранить копии объектов ингредиентов - лучше просто добавить к ним поле с количеством в корзинке.
Поправил код по комментариям

frrenzy commented:
Убрал стейт корзины, подумал, что незачем хранить копии объектов ингредиентов - лучше просто добавить к ним поле с количеством в корзинке.
Поправил код по комментариям

Zaytsev-Alex commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -75,54 +80,24 @@ const BurgerIngredients = ({ ingredients, cart, updateCart }) => {
-        )}
+        {[buns, sauces, mains].map(({ name, type, ingredients }, index) => (
+          <IngredientsSection
+            key={name}
```
Можно лучше: в случае компонента бургер ингредиентов наилучшим решением для key будет использование идентификатора ингредиента, который хранится в поле _id его объекта. Он будет явно уникальный и постоянный, в то время как название чисто теоретически может быть неуникальным.

frrenzy replied:
Ну тут же ключик именно секции - секции с булочками, секции с соусами и секции с начинками

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -8,44 +8,56 @@ import {
-            />
-          </li>
-        ))}
+        {ingredients.map(({ name, price, image, _id, count }, index) => {
```
Можно лучше: скролл в конструкторе бургер надо реализовать таким способом, чтобы скроллилась секция ингредиентов бургера, то, что находится между верхней и нижней булочками.
Проще всего: сделать это через 3 элемента: Компонент с булочкой, контейнер с основными ингредиентами, компонент с булочкой. И задать скролл только контейнеру с основными ингредиентами.

Zaytsev-Alex commented on src/components/header-button/header-button.jsx:
```diff
@@ -8,8 +8,9 @@ const HeaderButton = ({ text, value, isActive, onClick, icon: Icon }) => {
   return (
-    <button
+    <a
       type='button'
```
Можно лучше: type следует убрать у элемента ссылки.

frrenzy replied:
+

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -8,44 +8,56 @@ import {
-          </li>
-        ))}
+        {ingredients.map(({ name, price, image, _id, count }, index) => {
+          {
```
Можно лучше: фигурные скобочки этого блока можно убрать, они излишни.

frrenzy replied:
+

Zaytsev-Alex reviewed (commented):
Отличная работа над рефакторингом кода!

Из того, что можно сделать лучше, по результатам этой итерации:
- Использовать поле _id ингредиента в качестве key в компоненте ингредиентов бургера;
- Задать скролл в бургер конструкторе только для контейнера начинок бургера, чтобы обе булочки: верхняя и нижняя всегда были навиду;
- Убрать type у элементы ссылки.

Предварительная сдача окончена, этот ПР можно будет мержить. Далее необходимо реализовать функционал требуемый для финального этапа и прислать ссылку на ПР с решением через интерфейс сдачи, в который вы отправляли эту работу. После того, как финальный этап будет реализован и все "надо исправить", которые могут появится при его проверки, будут исправлены, работа будет принята.

Удачи в выполнении финального этапа!

frrenzy committed:
fix: review #2
